### PR TITLE
Generalize inheritance and composition of model classes

### PR DESCRIPTION
**Pull request by michaelosthege**
This PR refactors the model classes such ultimately code duplication is reduced.

The full reasoning behind this inheritance/composition design is [described in a new documentation page](https://github.com/JuBiotech/calibr8/blob/generalize-inheritance/docs/source/calibr8_inheritance.md).

## Changes
__Docs__ (done with #14)
* [x] Sphinx was configured to also render Markdown pages.
* [x] A Markdown document explaining the model classes, inheritance and distribution/noisemodel mixins was added.

__Utils__ (done with #14)
* [x] The `HAS_PYMC` constant is now exposed as `calibr8.HAS_PYMC`. And the `pm` module may be accessed via `calibr8.utils.pm`.

__Generalization of noise model__ (see #16)
* [x] Implement `calibr8.core.DistributionMixin` and a few `calibr8.*Noise` classes.
* [x] Update inheritance of existing contrib models to use mixins.
* [x] Refactor `BaseModelT.loglikelihood` to be distribution-agnostic.
* [x] Move `BaseModelT.loglikelihood` → `CalibrationModel.loglikelihood`.
* [x] Refactor `(log)likelihood` to slice correctly independent of `ndim(x)`.

__Inference type and dimensionality__
* [x] Split `InferenceResult` into subclasses.
* [x] Introduce `ContinuousUnivariateModel` and `ContinuousMultivariateModel` model subclasses with `infer_independent` implementations.
* [x] Update contrib models to inherit from model subclasses.

__Tests__ (see #17)
+ [x] Close multiple tests coverage gaps & fix some bugs in the test code.
+ [x] Reorganize the test suite to correspond to the class hierarchy.
+ Add x- & y-parameterized shape broadcasting tests for the log-likelihood implementation. (deferred to #15)

__Addition of examples__ (deferred to follow-up)
* Univariate, continuous with Laplace noise
* Univariate, discrete with Poisson noise
* Bivarite, continuous with Normal noise

### CODE REVIEW

File: docs/source/conf.py [hunk 1]
```diff
@@ -33,6 +33,7 @@
 extensions = [
     "sphinx.ext.autodoc",
     "sphinx_rtd_theme",
+    "myst_parser",
```
Comment (lhelleckes) [inline docs/source/conf.py:36]: What's that?

**Review by lhelleckes — COMMENTED** [review]
I think a good point to split would be the noise refactoring, which is a big enough change to stand for itself. I would for sure split the docs etc. because they need to be changed when everything else is done.

I added two comment here already, I hope that doesn't confuse you. For a detailed review, I would split it in three smaller PRs: 

Noise-related, InferenceResult-related and documentation-related changes.

**Comment by lhelleckes** [comment]
You also asked for weaknesses: except for the (probably) unnecessary subclass that I commented on above, I think we should clarify how the distributions would work in the discrete case. I think it's okay that this is not implemented for the first publication, but it should be clarified at least (that is also a nice outlook for the paper).

File: docs/source/conf.py (hunk 1 above)
Comment (michaelosthege) [reply docs/source/conf.py:36]: This is needed for rendering Markdown. I took this from the documentation. Link is in the commit message

File: docs/source/calibr8_inheritance.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+1. What kind of independent variable is the model working with? (continuous/discrete)
+2. How many dimensions does the independent variable have? (univariate/multivariate)
+
+Note that the `(log)likelihood` of a model is agnostic to the dependent variable being continuous/discrete, and therefore we do not have to consider it here.
+
+The following tables summarizes the long-form matrix of the various combinations and corresponding implementations.
+
+| ndim | independent variable | inference strategy | `InferenceResult` (properties) | model subclass |
+|----------------|----------------------|--------------------|--------------------------------|----------------|
+| 1 | continuous | `trapz` | `ContinuousUnivariateInference` (ETI, HDI, median) | `ContinuousUnivariateModel` |
```
Comment (lhelleckes) [inline docs/source/calibr8_inheritance.md:21]:
```suggestion
| 1 | continuous | `trapz`, MCMC | `ContinuousUnivariateInference` (ETI, HDI, median) | `ContinuousUnivariateModel` |
```

**Review by lhelleckes — APPROVED** [review]
OK

**Pull request merged**